Crash when no cameras are available on Android

Using the mobile_scanner package makes no camera Android devices not compatible. 

Here is the solution: https://developer.android.com/guide/topics/manifest/uses-permission-element

Well, apparently the plugin crashes when the device has no camera at all.

For example, if I disable front/back camera for the emulator, I get

```
W/CameraX ( 4715): androidx.camera.core.impl.CameraValidator$CameraIdListIncorrectException: Expected camera missing from device.
W/CameraX ( 4715):      at androidx.camera.core.impl.CameraValidator.validateCameras(CameraValidator.java:97)
W/CameraX ( 4715):      at androidx.camera.core.CameraX.lambda$initAndRetryRecursively$2$androidx-camera-core-CameraX(CameraX.java:342)
W/CameraX ( 4715):      at androidx.camera.core.CameraX$$ExternalSyntheticLambda3.run(Unknown Source:10)
W/CameraX ( 4715):      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
W/CameraX ( 4715):      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
W/CameraX ( 4715):      at java.lang.Thread.run(Thread.java:764)
W/CameraX ( 4715): Caused by: java.lang.IllegalArgumentException: No available camera can be found
W/CameraX ( 4715):      at androidx.camera.core.CameraSelector.select(CameraSelector.java:82)
W/CameraX ( 4715):      at androidx.camera.core.impl.CameraValidator.validateCameras(CameraValidator.java:83)
W/CameraX ( 4715):      ... 5 more
D/CameraValidator( 4715): Verifying camera lens facing on generic_x86_arm, lensFacingInteger: null
E/CameraValidator( 4715): Camera LensFacing verification failed, existing cameras: []
W/CameraX ( 4715): Retry init. Start time 93452 current time 94540
W/CameraX ( 4715): androidx.camera.core.impl.CameraValidator$CameraIdListIncorrectException: Expected camera missing from device.
W/CameraX ( 4715):      at androidx.camera.core.impl.CameraValidator.validateCameras(CameraValidator.java:97)
W/CameraX ( 4715):      at androidx.camera.core.CameraX.lambda$initAndRetryRecursively$2$androidx-camera-core-CameraX(CameraX.java:342)
W/CameraX ( 4715):      at androidx.camera.core.CameraX$$ExternalSyntheticLambda3.run(Unknown Source:10)
W/CameraX ( 4715):      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
W/CameraX ( 4715):      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
W/CameraX ( 4715):      at java.lang.Thread.run(Thread.java:764)
W/CameraX ( 4715): Caused by: java.lang.IllegalArgumentException: No available camera can be found
W/CameraX ( 4715):      at androidx.camera.core.CameraSelector.select(CameraSelector.java:82)
W/CameraX ( 4715):      at androidx.camera.core.impl.CameraValidator.validateCameras(CameraValidator.java:83)
W/CameraX ( 4715):      ... 5 more
D/CameraValidator( 4715): Verifying camera lens facing on generic_x86_arm, lensFacingInteger: null
E/CameraValidator( 4715): Camera LensFacing verification failed, existing cameras: []
W/CameraX ( 4715): Retry init. Start time 93452 current time 95082
W/CameraX ( 4715): androidx.camera.core.impl.CameraValidator$CameraIdListIncorrectException: Expected camera missing from device.
W/CameraX ( 4715):      at androidx.camera.core.impl.CameraValidator.validateCameras(CameraValidator.java:97)
W/CameraX ( 4715):      at androidx.camera.core.CameraX.lambda$initAndRetryRecursively$2$androidx-camera-core-CameraX(CameraX.java:342)
W/CameraX ( 4715):      at androidx.camera.core.CameraX$$ExternalSyntheticLambda3.run(Unknown Source:10)
W/CameraX ( 4715):      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
W/CameraX ( 4715):      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
W/CameraX ( 4715):      at java.lang.Thread.run(Thread.java:764)
W/CameraX ( 4715): Caused by: java.lang.IllegalArgumentException: No available camera can be found
W/CameraX ( 4715):      at androidx.camera.core.CameraSelector.select(CameraSelector.java:82)
W/CameraX ( 4715):      at androidx.camera.core.impl.CameraValidator.validateCameras(CameraValidator.java:83)
W/CameraX ( 4715):      ... 5 more
D/CameraValidator( 4715): Verifying camera lens facing on generic_x86_arm, lensFacingInteger: null
E/CameraValidator( 4715): Camera LensFacing verification failed, existing cameras: []
W/CameraX ( 4715): Retry init. Start time 93452 current time 95629
W/CameraX ( 4715): androidx.camera.core.impl.CameraValidator$CameraIdListIncorrectException: Expected camera missing from device.
W/CameraX ( 4715):      at androidx.camera.core.impl.CameraValidator.validateCameras(CameraValidator.java:97)
W/CameraX ( 4715):      at androidx.camera.core.CameraX.lambda$initAndRetryRecursively$2$androidx-camera-core-CameraX(CameraX.java:342)
W/CameraX ( 4715):      at androidx.camera.core.CameraX$$ExternalSyntheticLambda3.run(Unknown Source:10)
W/CameraX ( 4715):      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
W/CameraX ( 4715):      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
W/CameraX ( 4715):      at java.lang.Thread.run(Thread.java:764)
W/CameraX ( 4715): Caused by: java.lang.IllegalArgumentException: No available camera can be found
W/CameraX ( 4715):      at androidx.camera.core.CameraSelector.select(CameraSelector.java:82)
W/CameraX ( 4715):      at androidx.camera.core.impl.CameraValidator.validateCameras(CameraValidator.java:83)
W/CameraX ( 4715):      ... 5 more
D/CameraValidator( 4715): Verifying camera lens facing on generic_x86_arm, lensFacingInteger: null
E/CameraValidator( 4715): Camera LensFacing verification failed, existing cameras: []
E/CameraX ( 4715): The device might underreport the amount of the cameras. Finish the initialize task since we are already reaching the maximum number of retries.
D/AndroidRuntime( 4715): Shutting down VM
E/AndroidRuntime( 4715): FATAL EXCEPTION: main
E/AndroidRuntime( 4715): Process: dev.steenbakker.mobile_scanner_example, PID: 4715
E/AndroidRuntime( 4715): java.lang.IllegalArgumentException: Provided camera selector unable to resolve a camera for the given use case
E/AndroidRuntime( 4715):        at androidx.camera.lifecycle.ProcessCameraProvider.bindToLifecycle(ProcessCameraProvider.java:476)
E/AndroidRuntime( 4715):        at androidx.camera.lifecycle.ProcessCameraProvider.bindToLifecycle(ProcessCameraProvider.java:363)
E/AndroidRuntime( 4715):        at dev.steenbakker.mobile_scanner.MobileScanner.start$lambda$11(MobileScanner.kt:293)
E/AndroidRuntime( 4715):        at dev.steenbakker.mobile_scanner.MobileScanner.$r8$lambda$QBaqIq_loyYbiAjfq_M2yLd3JMk(Unknown Source:0)
E/AndroidRuntime( 4715):        at dev.steenbakker.mobile_scanner.MobileScanner$$ExternalSyntheticLambda1.run(Unknown Source:18)
E/AndroidRuntime( 4715):        at android.os.Handler.handleCallback(Handler.java:873)
E/AndroidRuntime( 4715):        at android.os.Handler.dispatchMessage(Handler.java:99)
E/AndroidRuntime( 4715):        at android.os.Looper.loop(Looper.java:193)
E/AndroidRuntime( 4715):        at android.app.ActivityThread.main(ActivityThread.java:6669)
E/AndroidRuntime( 4715):        at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime( 4715):        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
E/AndroidRuntime( 4715):        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
I/Process ( 4715): Sending signal. PID: 4715 SIG: 9
```

--- a/android/src/main/AndroidManifest.xml
+++ b/android/src/main/AndroidManifest.xml
@@ -2,4 +2,5 @@
   package="dev.steenbakker.mobile_scanner">
 
     <uses-permission android:name="android.permission.CAMERA" />
+    <uses-feature android:name="android.hardware.camera" android:required="false" />
 </manifest>

--- a/android/src/main/kotlin/dev/steenbakker/mobile_scanner/MobileScanner.kt
+++ b/android/src/main/kotlin/dev/steenbakker/mobile_scanner/MobileScanner.kt
@@ -235,10 +235,12 @@ class MobileScanner(
 
         cameraProviderFuture.addListener({
             cameraProvider = cameraProviderFuture.get()
+
             if (cameraProvider == null) {
                 throw CameraError()
             }
-            cameraProvider!!.unbindAll()
+
+            cameraProvider?.unbindAll()
             textureEntry = textureRegistry.createSurfaceTexture()
 
             // Preview
@@ -297,31 +299,34 @@ class MobileScanner(
                 analysis
             )
 
-            // Register the torch listener
-            camera!!.cameraInfo.torchState.observe(activity) { state ->
-                // TorchState.OFF = 0; TorchState.ON = 1
-                torchStateCallback(state)
+            camera?.let {
+                // Register the torch listener
+                it.cameraInfo.torchState.observe(activity as LifecycleOwner) { state ->
+                    // TorchState.OFF = 0; TorchState.ON = 1
+                    torchStateCallback(state)
+                }
+
+                // Register the zoom scale listener
+                it.cameraInfo.zoomState.observe(activity) { state ->
+                    zoomScaleStateCallback(state.linearZoom.toDouble())
+                }
+
+                // Enable torch if provided
+                if (it.cameraInfo.hasFlashUnit()) {
+                    it.cameraControl.enableTorch(torch)
+                }
             }
-
-            // Register the zoom scale listener
-            camera!!.cameraInfo.zoomState.observe(activity) { state ->
-                zoomScaleStateCallback(state.linearZoom.toDouble())
-            }
-
-
-            // Enable torch if provided
-            camera!!.cameraControl.enableTorch(torch)
 
             val resolution = analysis.resolutionInfo!!.resolution
-            val portrait = camera!!.cameraInfo.sensorRotationDegrees % 180 == 0
             val width = resolution.width.toDouble()
             val height = resolution.height.toDouble()
+            val portrait = (camera?.cameraInfo?.sensorRotationDegrees ?: 0) % 180 == 0
 
             mobileScannerStartedCallback(
                 MobileScannerStartParameters(
                     if (portrait) width else height,
                     if (portrait) height else width,
-                    camera!!.cameraInfo.hasFlashUnit(),
+                    camera?.cameraInfo?.hasFlashUnit() ?: false,
                     textureEntry!!.id()
                 )
             )
@@ -363,7 +368,10 @@ class MobileScanner(
         if (camera == null) {
             throw TorchWhenStopped()
         }
-        camera!!.cameraControl.enableTorch(enableTorch)
+
+        if (camera?.cameraInfo?.hasFlashUnit() == true) {
+            camera?.cameraControl?.enableTorch(enableTorch)
+        }
     }
 
     /**
@@ -393,9 +401,9 @@ class MobileScanner(
      * Set the zoom rate of the camera.
      */
     fun setScale(scale: Double) {
-        if (camera == null) throw ZoomWhenStopped()
         if (scale > 1.0 || scale < 0) throw ZoomNotInRange()
-        camera!!.cameraControl.setLinearZoom(scale.toFloat())
+        if (camera == null) throw ZoomWhenStopped()
+        camera?.cameraControl?.setLinearZoom(scale.toFloat())
     }
 
     /**
@@ -403,7 +411,7 @@ class MobileScanner(
      */
     fun resetScale() {
         if (camera == null) throw ZoomWhenStopped()
-        camera!!.cameraControl.setZoomRatio(1f)
+        camera?.cameraControl?.setZoomRatio(1f)
     }
 
 }

--- a/android/src/main/kotlin/dev/steenbakker/mobile_scanner/MobileScanner.kt
+++ b/android/src/main/kotlin/dev/steenbakker/mobile_scanner/MobileScanner.kt
@@ -28,7 +28,6 @@ import android.view.WindowManager
 import android.content.Context
 import android.os.Build
 
-
 class MobileScanner(
     private val activity: Activity,
     private val textureRegistry: TextureRegistry,
@@ -213,6 +212,7 @@ class MobileScanner(
         torchStateCallback: TorchStateCallback,
         zoomScaleStateCallback: ZoomScaleStateCallback,
         mobileScannerStartedCallback: MobileScannerStartedCallback,
+        mobileScannerErrorCallback: (exception: Exception) -> Unit,
         detectionTimeout: Long,
         cameraResolution: Size?
     ) {
@@ -221,7 +221,9 @@ class MobileScanner(
         this.returnImage = returnImage
 
         if (camera?.cameraInfo != null && preview != null && textureEntry != null) {
-            throw AlreadyStarted()
+            mobileScannerErrorCallback(AlreadyStarted())
+
+            return
         }
 
         scanner = if (barcodeScannerOptions != null) {
@@ -237,7 +239,9 @@ class MobileScanner(
             cameraProvider = cameraProviderFuture.get()
 
             if (cameraProvider == null) {
-                throw CameraError()
+                mobileScannerErrorCallback(CameraError())
+
+                return@addListener
             }
 
             cameraProvider?.unbindAll()
@@ -292,12 +296,18 @@ class MobileScanner(
 
             val analysis = analysisBuilder.build().apply { setAnalyzer(executor, captureOutput) }
 
-            camera = cameraProvider!!.bindToLifecycle(
-                activity as LifecycleOwner,
-                cameraPosition,
-                preview,
-                analysis
-            )
+            try {
+                camera = cameraProvider?.bindToLifecycle(
+                    activity as LifecycleOwner,
+                    cameraPosition,
+                    preview,
+                    analysis
+                )
+            } catch(exception: Exception) {
+                mobileScannerErrorCallback(NoCamera())
+
+                return@addListener
+            }
 
             camera?.let {
                 // Register the torch listener

--- a/android/src/main/kotlin/dev/steenbakker/mobile_scanner/MobileScannerExceptions.kt
+++ b/android/src/main/kotlin/dev/steenbakker/mobile_scanner/MobileScannerExceptions.kt
@@ -3,7 +3,6 @@ package dev.steenbakker.mobile_scanner
 class NoCamera : Exception()
 class AlreadyStarted : Exception()
 class AlreadyStopped : Exception()
-class TorchError : Exception()
 class CameraError : Exception()
 class TorchWhenStopped : Exception()
 class ZoomWhenStopped : Exception()

--- a/android/src/main/kotlin/dev/steenbakker/mobile_scanner/MobileScannerHandler.kt
+++ b/android/src/main/kotlin/dev/steenbakker/mobile_scanner/MobileScannerHandler.kt
@@ -238,7 +238,7 @@ class MobileScannerHandler(
         try {
             mobileScanner!!.toggleTorch(call.arguments == 1)
             result.success(null)
-        } catch (e: AlreadyStopped) {
+        } catch (e: TorchWhenStopped) {
             result.error("MobileScanner", "Called toggleTorch() while stopped!", null)
         }
     }

--- a/android/src/main/kotlin/dev/steenbakker/mobile_scanner/MobileScannerHandler.kt
+++ b/android/src/main/kotlin/dev/steenbakker/mobile_scanner/MobileScannerHandler.kt
@@ -165,58 +165,60 @@ class MobileScannerHandler(
 
         val detectionSpeed: DetectionSpeed = DetectionSpeed.values().first { it.intValue == speed}
 
-        try {
-            mobileScanner!!.start(
-                barcodeScannerOptions,
-                returnImage, 
-                position,
-                torch,
-                detectionSpeed,
-                torchStateCallback,
-                zoomScaleStateCallback,
-                mobileScannerStartedCallback = {
-                    Handler(Looper.getMainLooper()).post {
-                        result.success(mapOf(
-                            "textureId" to it.id,
-                            "size" to mapOf("width" to it.width, "height" to it.height),
-                            "torchable" to it.hasFlashUnit
-                        ))
+        mobileScanner!!.start(
+            barcodeScannerOptions,
+            returnImage,
+            position,
+            torch,
+            detectionSpeed,
+            torchStateCallback,
+            zoomScaleStateCallback,
+            mobileScannerStartedCallback = {
+                Handler(Looper.getMainLooper()).post {
+                    result.success(mapOf(
+                        "textureId" to it.id,
+                        "size" to mapOf("width" to it.width, "height" to it.height),
+                        "torchable" to it.hasFlashUnit
+                    ))
+                }
+            },
+            mobileScannerErrorCallback = {
+                Handler(Looper.getMainLooper()).post {
+                    when (it) {
+                        is AlreadyStarted -> {
+                            result.error(
+                                "MobileScanner",
+                                "Called start() while already started",
+                                null
+                            )
+                        }
+                        is CameraError -> {
+                            result.error(
+                                "MobileScanner",
+                                "Error occurred when setting up camera!",
+                                null
+                            )
+                        }
+                        is NoCamera -> {
+                            result.error(
+                                "MobileScanner",
+                                "No camera found or failed to open camera!",
+                                null
+                            )
+                        }
+                        else -> {
+                            result.error(
+                                "MobileScanner",
+                                "Unknown error occurred.",
+                                null
+                            )
+                        }
                     }
-                },
-                timeout.toLong(),
-                cameraResolution,
-            )
-        } catch (e: AlreadyStarted) {
-            result.error(
-                "MobileScanner",
-                "Called start() while already started",
-                null
-            )
-        } catch (e: NoCamera) {
-            result.error(
-                "MobileScanner",
-                "No camera found or failed to open camera!",
-                null
-            )
-        } catch (e: TorchError) {
-            result.error(
-                "MobileScanner",
-                "Error occurred when setting torch!",
-                null
-            )
-        } catch (e: CameraError) {
-            result.error(
-                "MobileScanner",
-                "Error occurred when setting up camera!",
-                null
-            )
-        } catch (e: Exception) {
-            result.error(
-                "MobileScanner",
-                "Unknown error occurred..",
-                null
-            )
-        }
+                }
+            },
+            timeout.toLong(),
+            cameraResolution,
+        )
     }
 
     private fun stop(result: MethodChannel.Result) {

--- a/ios/Classes/MobileScanner.swift
+++ b/ios/Classes/MobileScanner.swift
@@ -305,7 +305,7 @@ public class MobileScanner: NSObject, AVCaptureVideoDataOutputSampleBufferDelega
     /// Set the zoom factor of the camera
     func setScale(_ scale: CGFloat) throws {
         if (device == nil) {
-            throw MobileScannerError.torchWhenStopped
+            throw MobileScannerError.zoomWhenStopped
         }
         
         do {

--- a/ios/Classes/MobileScannerPlugin.swift
+++ b/ios/Classes/MobileScannerPlugin.swift
@@ -95,7 +95,7 @@ public class MobileScannerPlugin: NSObject, FlutterPlugin {
         }
     }
 
-    /// Parses all parameters and starts the mobileScanner
+    /// Start the mobileScanner.
     private func start(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
         let torch: Bool = (call.arguments as! Dictionary<String, Any?>)["torch"] as? Bool ?? false
         let facing: Int = (call.arguments as! Dictionary<String, Any?>)["facing"] as? Int ?? 1
@@ -151,7 +151,7 @@ public class MobileScannerPlugin: NSObject, FlutterPlugin {
         }
     }
 
-    /// Stops the mobileScanner and closes the texture
+    /// Stops the mobileScanner and closes the texture.
     private func stop(_ result: @escaping FlutterResult) {
         do {
             try mobileScanner.stop()
@@ -159,7 +159,7 @@ public class MobileScannerPlugin: NSObject, FlutterPlugin {
         result(nil)
     }
 
-    /// Toggles the torch
+    /// Toggles the torch.
     private func toggleTorch(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
         do {
             try mobileScanner.toggleTorch(call.arguments as? Int == 1 ? .on : .off)
@@ -171,7 +171,7 @@ public class MobileScannerPlugin: NSObject, FlutterPlugin {
         }
     }
     
-    /// Toggles the zoomScale
+    /// Sets the zoomScale.
     private func setScale(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
         let scale = call.arguments as? CGFloat
         if (scale == nil) {
@@ -198,7 +198,7 @@ public class MobileScannerPlugin: NSObject, FlutterPlugin {
         }
     }
 
-    /// Reset the zoomScale
+    /// Reset the zoomScale.
     private func resetScale(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
         do {
             try mobileScanner.resetScale()
@@ -218,7 +218,7 @@ public class MobileScannerPlugin: NSObject, FlutterPlugin {
         }
     }
 
-    /// Toggles the torch
+    /// Updates the scan window rectangle.
     func updateScanWindow(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
         let scanWindowData: Array? = (call.arguments as? [String: Any])?["rect"] as? [CGFloat]
         MobileScannerPlugin.scanWindow = scanWindowData
@@ -240,7 +240,7 @@ public class MobileScannerPlugin: NSObject, FlutterPlugin {
         return CGRect(x: minX, y: minY, width: width, height: height)
     }
     
-    /// Analyzes a single image
+    /// Analyzes a single image.
     private func analyzeImage(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
         let uiImage = UIImage(contentsOfFile: call.arguments as? String ?? "")
         

